src/OAuth/ResourceOwner/TelegramResourceOwner.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/*
+ * This file is part of the HWIOAuthBundle package.
+ *
+ * (c) Hardware Info <opensource@hardware.info>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace HWI\Bundle\OAuthBundle\OAuth\ResourceOwner;
+
+use HWI\Bundle\OAuthBundle\OAuth\Response\PathUserResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\LazyResponseException;
+
+/**
+ * @author zorn-v
+ */
+class TelegramResourceOwner extends GenericOAuth2ResourceOwner
+{
+    public const TYPE = 'telegram';
+
+    protected array $paths = [
+        'identifier' => 'id',
+        'nickname' => 'username',
+        'realname' => 'first_name',
+        'firstname' => 'first_name',
+        'lastname' => 'last_name',
+        'profilepicture' => 'photo_url',
+    ];
+
+    public function getAuthorizationUrl($redirectUri, array $extraParameters = [])
+    {
+        list($botId) = explode(':', $this->options['client_secret']);
+        $parameters = array_merge([
+            'bot_id' => $botId,
+            'origin' => $redirectUri,
+            'return_to' => $redirectUri,
+        ], $extraParameters);
+
+        return $this->normalizeUrl($this->options['authorization_url'], $parameters);
+    }
+
+    public function handles(Request $request)
+    {
+        if (!$request->query->has('code')) {
+            $js = sprintf('<script>location.href = "?code=" + new URLSearchParams(location.hash.substring(1)).get("tgAuthResult")</script>');
+            throw new LazyResponseException(new Response($js));
+        }
+        return true;
+    }
+
+    public function getAccessToken(Request $request, $redirectUri, array $extraParameters = [])
+    {
+        $token = $request->query->get('code');
+        $data = json_decode(base64_decode($token), true);
+        if (!$data) {
+            throw new AuthenticationException('Missing Telegram data');
+        }
```

Comment:
I would add a throw exception on invalid and catch it + pack with that exception, cause now it hides potential invalid data from Telegram and says they are missing instead which is not true.

Reply:
If data can not be decoded then it is not from telegram, no ? 
Maybe `Invalid Telegram data` then ?

Reply:
Also it is probably base64url encoded (as it is in url), so I'll add proper decoding.